### PR TITLE
July 2024 Release

### LINKED ISSUE #10: July 2024 Release
- [x] Added SAC
- [x] Inference model saving based on average reward increase
- [x] Small bug fixes
- [x] Optimizations
- [x] Inference Actions on Tensorboard

### PR DESCRIPTION
**Pull request by armenkasp**
# Overall Additions

- [x] Added SAC
- [x] Inference model saving based on average reward increase
- [x] Small bug fixes
- [x] Optimizations
- [x] Inference Actions on Tensorboard

# Benchmark Results:

Screenshot 2024-07-19 at 4 11 35 PM
Screenshot 2024-07-19 at 4 11 44 PM
Screenshot 2024-07-19 at 4 11 25 PM

# Unit Tests Results:

Screenshot 2024-07-19 at 2 22 53 PM
Screenshot 2024-07-19 at 3 23 41 PM

# New Tensorboard Action Inference Output

image

### CODE REVIEW

File: jlab_opt_control/agents/keras_sac.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+            sampled_action = sampled_action.numpy().flatten()
+            noise = noise.numpy().flatten()
+
+
+        # Log the training action(s) taken
+        if train:
+            self.nactions = self.nactions + 1
+            if self.num_actions == 0:
+                tf.summary.scalar('Action', data=sampled_action,
+                                  step=int(self.nactions))
```
Comment (Kishanrajput) [inline jlab_opt_control/agents/keras_sac.py]: I'm not sure I follow, how can num_actions be 0?

Comment (Kishanrajput) [reply jlab_opt_control/agents/keras_sac.py]: Should it be 
`self.num_actions == 1:` ?

File: jlab_opt_control/agents/keras_ddpg.py [hunk 2]
```diff
@@ -308,7 +309,15 @@ def action(self, state, train=True, inference=False):
                 for i in range(self.num_actions):
                     tf.summary.scalar('Action #{}'.format(
                         i), data=sampled_action[i], step=int(self.nactions))
-
+        if inference:
+            self.inf_nactions = self.inf_nactions + 1
+            if self.num_actions == 0:
```
Comment (Kishanrajput) [inline jlab_opt_control/agents/keras_ddpg.py]: Same comment as SAC, `self.num_actions == 0:` --> should it be `self.num_actions == 1:`?

File: jlab_opt_control/agents/keras_td3.py [hunk 3]
```diff
@@ -349,6 +349,15 @@ def action(self, state, train=True, inference=False):
                 for i in range(self.num_actions):
                     tf.summary.scalar('Action #{}'.format(
                         i), data=sampled_action[i], step=int(self.nactions))
+        if inference:
+            self.inf_nactions = self.inf_nactions + 1
+            if self.num_actions == 0:
```
Comment (Kishanrajput) [inline jlab_opt_control/agents/keras_td3.py]: `self.num_actions == 0:` --> should it be `self.num_actions == 1:`?

**Review by Kishanrajput — CHANGES_REQUESTED** [review]
I think it looks good overall. Please have a look at the comments below.

File: jlab_opt_control/agents/keras_td3.py (hunk 3 above)
Comment (armenkasp) [reply jlab_opt_control/agents/keras_td3.py]: Changed the code in the agents to not be linked to size of action space for tensorboard logging!

**Review by Kishanrajput — APPROVED** [review]
Thank you for addressing my comments. I ran the "tests" and everything pass. However, I do not think we have proper unit tests, perhaps something to focus in the next release.

**Pull request merged**